### PR TITLE
cc-wrapper: fix wrapper tests and building for darwin on staging-next

### PR DESCRIPTION
**Pull request by Emantor**
###### Motivation for this change
Fix darwin builds and the cc-wrapper-libcxx tests for staging-next. Opening as draft since I am not sure the first commit is the correct fix for this.

###### Things done

- [x] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [ ] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

### CODE REVIEW

File: pkgs/build-support/cc-wrapper/default.nix [hunk 1]
```diff
@@ -154,6 +153,7 @@ stdenv.mkDerivation {
         local dst="$1"
         local wrapper="$2"
         export prog="$3"
+        export isdarwin="${toString targetPlatform.isDarwin or false}"
```
Comment (Emantor) [inline pkgs/build-support/cc-wrapper/default.nix:156]: as I almost expected, this does not work, the substitution is not performed in the wrapper script.

**Review by LnL7 — COMMENTED** [review]
Polluting the wrappers with this type of logic doesn't seem ideal. Allowing this to be set in the environment and inheriting that from the stdenv seems like a nicer approach. Some similar things are done like that here.

https://github.com/NixOS/nixpkgs/blob/8946ff8d71313fbe9634e242149f1e3fb6aa824b/pkgs/stdenv/darwin/default.nix#L36-L38

**Comment by Emantor** [comment]
> Polluting the wrappers with this type of logic doesn't seem ideal. Allowing this to be set in the environment and inheriting that from the stdenv seems like a nicer approach. Some similar things are done like that here.
> 
> https://github.com/NixOS/nixpkgs/blob/8946ff8d71313fbe9634e242149f1e3fb6aa824b/pkgs/stdenv/darwin/default.nix#L36-L38

I strongly agree, that is why I opened it as a draft. What do you think is the correct approach here, setting `NIX_DYNAMIC_LINKER` to `darwin` and checking for that? Or do we need another variable for clarity?

**Comment by Ericson2314** [comment]
CC @Mic92. I'm curious what exactly want wrong with #91293/#92089/#94582 and #106908.

#106925 in particular was supposed to help with this exact issue?

**Comment by LnL7** [comment]
I would change `setDynamicLinker=1` to `NIX_SET_DYNAMIC_LINKER=${NIX_SET_DYNAMIC_LINKER:-1}`.

@Ericson2314 That's just the test, the stdenv is still broken.

**Comment by Emantor** [comment]
> I would change `setDynamicLinker=1` to `NIX_SET_DYNAMIC_LINKER=${NIX_SET_DYNAMIC_LINKER:-1}`.

Okay, I just realized that my first fixup was also broken… it's late here so I'll be going to bed, if this hasn't been resolved by tomorrow I'll work on this some more.

**Comment by Emantor** [comment]
Closing this, since the two changes have been reverted in staging-next.

**Pull request closed without merging**